### PR TITLE
Bug 1949021: UPSTREAM: 101422: rbac-authz: return "informers not synced" reason

### PR DESCRIPTION
**Pull request by sttts**
Backport of https://github.com/kubernetes/kubernetes/pull/101422.

### CODE REVIEW

File: pkg/kubeapiserver/authorizer/config.go [hunk 1]
```diff
@@ -153,3 +170,37 @@ func (config Config) New() (authorizer.Authorizer, authorizer.RuleResolver, erro
...
+
+func (az withNotSyncedReason) Authorize(ctx context.Context, a authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+	d, reason, err := az.ruleResolvedAUthorizer.Authorize(ctx, a)
+	if err != nil {
+		return d, reason, err
+	}
+
+	if d == authorizer.DecisionNoOpinion && len(reason) == 0 && !az.hasSynced() {
+		reason = "informers not synced"
+		klog.Warningf("authorize before synced informers by %s", a.GetUser().GetName())
```
Comment (sttts) [inline pkg/kubeapiserver/authorizer/config.go:192]: note that this should never happen, only if we get request early or late.

File: pkg/kubeapiserver/authorizer/config.go [hunk 2]
```diff
@@ -95,7 +98,15 @@ func (config Config) New() (authorizer.Authorizer, authorizer.RuleResolver, erro
...
 			)
-			nodeAuthorizer := node.NewAuthorizer(graph, nodeidentifier.NewDefaultNodeIdentifier(), bootstrappolicy.NodeRules())
+			nodeAuthorizer := withNotSyncedReason{
+				node.NewAuthorizer(graph, nodeidentifier.NewDefaultNodeIdentifier(), bootstrappolicy.NodeRules()),
+				[]cache.SharedInformer{
+					config.VersionedInformerFactory.Core().V1().Nodes().Informer(),
+					config.VersionedInformerFactory.Core().V1().Pods().Informer(),
+					config.VersionedInformerFactory.Core().V1().PersistentVolumes().Informer(),
+					config.VersionedInformerFactory.Storage().V1().VolumeAttachments().Informer(),
+				},
```
Comment (stlaz) [inline pkg/kubeapiserver/authorizer/config.go:108]: how do we make sure the list of informers is consistent with what's used inside?

**Review by stlaz — COMMENTED** [review]
why is empty reason bad on "NoOpinion"?

File: pkg/kubeapiserver/authorizer/config.go (hunk 2 above)
Comment (sttts) [reply pkg/kubeapiserver/authorizer/config.go:108]: just manually by looking 10 lines above. But this is the case everywhere where we have similar code. Better pattern welcome.

**Comment by sttts** [comment]
@stlaz 

```
W0414 08:32:21.299510      19 config.go:192] authorize before synced informers by system:serviceaccount:openshift-kube-apiserver:check-endpoints
W0414 08:32:21.306574      19 config.go:192] authorize before synced informers by system:anonymous
```

File: pkg/kubeapiserver/authorizer/config.go [hunk 3]
```diff
@@ -153,3 +170,37 @@ func (config Config) New() (authorizer.Authorizer, authorizer.RuleResolver, erro
...
+
+func (az withNotSyncedReason) Authorize(ctx context.Context, a authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+	d, reason, err := az.ruleResolvedAUthorizer.Authorize(ctx, a)
+	if err != nil {
+		return d, reason, err
+	}
+
+	if d == authorizer.DecisionNoOpinion && len(reason) == 0 && !az.hasSynced() {
+		reason = "informers not synced"
+		klog.Warningf("authorize before synced informers by %s for %s at %s", a.GetUser().GetName(), a.GetVerb(), a.GetPath())
```
Comment (stlaz) [inline pkg/kubeapiserver/authorizer/config.go:192]: format expects 3 arguments, got 4

Comment (sttts) [reply pkg/kubeapiserver/authorizer/config.go:192]: read precisely: there are only 3 :)

**Comment by smarterclayton** [comment]
```
Apr 23 12:08:54.000 W ns/e2e-test-router-idling-j57fl pod/idle-test-f5fb59756-s6hpk node/ci-op-qmg0lq0t-b2c4f-zb9cj-worker-b-f74vh reason/Failed Failed to pull image "image-registry.openshift-image-registry.svc:5000/openshift/tools:latest": rpc error: code = Unknown desc = Error reading manifest latest in image-registry.openshift-image-registry.svc:5000/openshift/tools: unauthorized: authentication required
```

in https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_kubernetes/666/pull-ci-openshift-kubernetes-master-e2e-gcp/1385547086795116544

Hrm... 

```
openshift-image-registry_image-registry-8497489544-x5rvx_registry.log:time="2021-04-23T12:08:54.137784227Z" level=error msg="OpenShift access denied: no opinion" go.version=go1.15.7 http.request.host="image-registry.openshift-image-registry.svc:5000" http.request.id=418b20fa-abb4-4e14-bc39-c84baba0b50f 
```

@dmage does this mean the error we're seeing isn't informer based?

**Comment by dmage** [comment]
@smarterclayton no, it isn't, if I understand this PR correctly. The registry [SAR status printer is simple](https://github.com/openshift/image-registry/blob/1b129a6c313f89e11d97601808ad7bc5b75d6009/pkg/dockerregistry/server/auth.go#L390-L413) and `"... access denied: no opinion"` means that Reason and EvaluationError are both empty.

File: pkg/kubeapiserver/authorizer/config.go [hunk 4]
```diff
@@ -153,3 +170,37 @@ func (config Config) New() (authorizer.Authorizer, authorizer.RuleResolver, erro
...
+	informers []cache.SharedInformer
+}
+
+func (az withNotSyncedReason) Authorize(ctx context.Context, a authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+	d, reason, err := az.ruleResolvedAUthorizer.Authorize(ctx, a)
+	if err != nil {
+		return d, reason, err
+	}
+
+	if d == authorizer.DecisionNoOpinion && len(reason) == 0 && !az.hasSynced() {
```
Comment (dmage) [inline pkg/kubeapiserver/authorizer/config.go:190]: Should we check `hasSynced` before calling `Authorize`? There might be a race if informers got synched right after we got results from Authorize.

File: pkg/kubeapiserver/authorizer/config.go [hunk 5]
```diff
@@ -153,3 +170,37 @@ func (config Config) New() (authorizer.Authorizer, authorizer.RuleResolver, erro
...
+		reason = "informers not synced"
```
Comment (dmage) [inline pkg/kubeapiserver/authorizer/config.go:191]: `InformersNotSynced`?

File: pkg/kubeapiserver/authorizer/config.go [hunk 6]
```diff
@@ -153,3 +170,37 @@ func (config Config) New() (authorizer.Authorizer, authorizer.RuleResolver, erro
 
 	return union.New(authorizers...), union.NewRuleResolvers(ruleResolvers...), nil
 }
+
+type ruleResolvedAUthorizer interface {
```
Comment (dmage) [inline pkg/kubeapiserver/authorizer/config.go:174]:
```suggestion
type ruleResolvedAuthorizer interface {
```

File: pkg/kubeapiserver/authorizer/config.go (hunk 4 above)
Comment (sttts) [reply pkg/kubeapiserver/authorizer/config.go:190]: done

File: pkg/kubeapiserver/authorizer/config.go (hunk 5 above)
Comment (sttts) [reply pkg/kubeapiserver/authorizer/config.go:191]: done

File: pkg/kubeapiserver/authorizer/config.go (hunk 6 above)
Comment (sttts) [reply pkg/kubeapiserver/authorizer/config.go:174]: done

**Comment by s-urbaniak** [comment]
@sttts it seems you have tagged @dmage's comments as done, but i think you didn't push the commits. do you mind to update the PR?

**Pull request closed without merging**